### ISSUE TITLE
Replace `Continuation_already_taken` exception with a predefined `Invalid_argument` exception

### ISSUE THREAD

**Issue by kayceesrk** [ocaml/ocaml#11422]
When a continuation is attempted to be resumed more than once, we get a `Continuation_already_taken` exception:

```ocaml
open Effect
open Effect.Deep

type _ Effect.t += E : unit t

let _ =
  try_with perform E
  { effc = fun (type a) (eff : a t) ->
      match eff with
      | E -> Some (fun (k : (a,_) continuation) ->
          continue k ();
          continue k ()) (* Continuation_already_taken exception raised here *)
      | e -> None }
```

`Continuation_already_taken` is a pre-defined exception, and was introduced to ensure that in the fast path, `continue` and `discontinue` functions do not go through the slower allocating C call (see PR https://github.com/ocaml-multicore/ocaml-multicore/pull/437). Before https://github.com/ocaml-multicore/ocaml-multicore/pull/437, the second continue will have raised `Invalid_argument "continuation_already_taken"` exception. Creating this exception requires an allocation, which this more specialised `Continuation_already_taken` exception avoids. Unfortunately, `Continuation_already_taken` needs to be a predefined exception since the [assembly stubs refer to it](https://github.com/ocaml/ocaml/blob/4b5dc163344ff03b50cd3d3fb202774ea992cdee/runtime/amd64.S#L947). It seems less hygenic to define a new predefined exception to avoid allocation.

## Proposal

An alternative proposal would be to have a global root which defines `Invalid_argument "continuation_already_taken"` once and for all and use that instead of `Continuation_already_taken`. This came up since I've been looking at the API around exceptions in Effect module in https://github.com/ocaml/ocaml/pull/11419.

**Comment by gasche** [comment]
I have a naive question about https://github.com/ocaml-multicore/ocaml-multicore/pull/437. From the patch I can see that `_noexc` verson of the primitive is added that neither allocate nor raises -- it returns `NULL` if an error happened -- and then a wrapper raises on error.

Question: what's the issue if the wrapper allocates? It should not affect the performance / calling conventions of the `_noexc` version, which I understood to be the fast path. Is it about the performance of the wrapper? (Is it faster to "possibly raise but never allocate" than to "possibly raise and allocate"? I thought that `noalloc` was not allowed either on functions that raise?)

**Comment by kayceesrk** [comment]
> what's the issue if the wrapper allocates?

It is not about allocations, but from where the exception gets raised. 

The idea in https://github.com/ocaml-multicore/ocaml-multicore/pull/437 is to avoid the slower "possibly-allocates-or-raises" c call (the one through `caml_c_call`) in the fast path when continuations are resumed successfully. Before https://github.com/ocaml-multicore/ocaml-multicore/pull/437, `caml_continuation_use` raised the `Continuation_already_taken` exception, which necessitated the slower `caml_c_call` path. Note that `caml_continuation_use` does not allocate but may raise an exception. With the `_noexc` version, the external call from OCaml to C goes through the fast path, 

https://github.com/ocaml/ocaml/blob/920fb8e99efd0bb18131fb22adb826b27defd197/asmcomp/amd64/emit.mlp#L556-L566

returning `NULL` if the error occurred. Then, the exception itself is raised by the wrapper `caml_resume`. Note that `caml_resume` also does not allocate. If the `caml_resume` or any other wrapper were to allocate, then it has to go through the slower path `caml_c_call`.

**Comment by xavierleroy** [comment]
> Then, the exception itself is raised by the wrapper caml_resume. Note that caml_resume also does not allocate. If the caml_resume or any other wrapper were to allocate, then it has to go through the slower path caml_c_call.

"Noalloc" C functions are not supposed to raise exceptions.  In other words, it's unsafe to raise an exception from C without going through `caml_c_call`.

For one thing, some ports cache the pointer to the current exception handler in a register (e.g. ARM64 using the x27 register).  `caml_c_call` stores the register in the `exn_handler` field of the domain state, so that the C function knows the current handler.  If you skip `caml_c_call`, the exception raised from C will jump to the wrong handler!

For another thing, exception raising from C can trigger async event processing which can trigger GC, so there must be valid frame descriptor associated with the call site, which is not the case for "noalloc" calls.  For the same reason, exception backtraces will be wrong too.

This is all written in the fine manual, https://v2.ocaml.org/releases/4.14/htmlman/intfc.html#ss:c-direct-call

> For small functions that are called repeatedly, this indirection can have a big impact on performances. However this is not needed if we know that the C function doesn’t allocate, doesn’t raise exceptions, and doesn’t release the master lock (see section [20.12.2](https://v2.ocaml.org/releases/4.14/htmlman/intfc.html#ss%3Aparallel-execution-long-running-c-code)). We can instruct the OCaml native-code compiler of this fact by annotating the external declaration with the attribute [@@noalloc]:

**Comment by favonia** [comment]
If we wish to use `Invalid_argument`, I slightly prefer one of the following alternatives:
```ocaml
Invalid_argument "continue/discontinue: continuation already taken"
Invalid_argument "continue/discontinue: continuation already resumed" (* if we rename the exception *)
Invalid_argument "Effect: continuation already resumed"
```
because in the current OCaml codebase, almost all instances I know fit into one of the following patterns:
```ocaml
Invalid_argument "function_name"
Invalid_argument "module/function_names: some explanation in plain English"
Invalid_argument "some explanation in plain English"
```
and it seems rare to use an argument that is neither a plain English phrase nor a function name.

**Comment by kayceesrk** [comment]
The motivation for doing this change is to avoid introducing the predefined `Continuation_already_taken` exception. Thanks to #11423 with which `Continuation_already_taken` no longer needs to be pre-defined. Hence, I do not propose to make this change. 

Wonder if it is useful to keep this issue open until at least #11423 is merged?

**Comment by xavierleroy** [comment]
Closing, as #11423 is about to be merged.
